src/Microsoft.DotNet.Build.VSTSBuildsApi/VSTSBuildClient.cs

```diff
@@ -0,0 +1,228 @@
+            RemoveJObjectToken("id", definition);
+
+            string requestUri = $"DefaultCollection/DevDiv/_apis/build/definitions?api-version=2.0";
+
+            HttpContent content = new StringContent(JsonConvert.SerializeObject(definition), Encoding.UTF8, "application/json");
+            ProcessResponseStatusCode(GetClient(definition["url"].ToString()).PostAsync(requestUri, content).Result);
+        }
+
+        /// <summary>
+        /// Retrieve a VSTS Build Definition by name and VSTS folder path
+        /// </summary>
+        /// <param name="name"></param>
+        /// <returns></returns>
+        private JObject VSTS_RetrieveDefinitionById(JObject definition)
+        {
+            string requestUri = $"DefaultCollection/DevDiv/_apis/build/definitions/{definition["id"].ToString()}?api-version=2.0";
+
+            HttpResponseMessage response = GetClient(definition["url"].ToString()).GetAsync(requestUri).Result;
+            ProcessResponseStatusCode(response);
+            return JObject.Parse(response.Content.ReadAsStringAsync().Result);
+        }
+
+        /// <summary>
+        /// Retrieve a VSTS Build Definition by name and VSTS folder path
+        /// </summary>
+        /// <param name="name"></param>
+        /// <returns></returns>
+        private IReadOnlyList<JObject> VSTS_RetrieveDefinitionsListByNameAndPath(JObject definition)
+        {
+            string requestUri = $"DefaultCollection/DevDiv/_apis/build/definitions?api-version=2.0&name={definition["name"].ToString()}";
+
+            HttpResponseMessage response = GetClient(definition["url"].ToString()).GetAsync(requestUri).Result;
+            ProcessResponseStatusCode(response);
+            string json = response.Content.ReadAsStringAsync().Result;
+            JObject definitionsJObject = JObject.Parse(json);
+            List<JObject> definitions = new List<JObject>();
+            if (int.Parse(definitionsJObject["count"].Value<string>()) > 0)
+            {
+                var values = definitionsJObject["value"].Children();
+                foreach(var value in values)
+                {
+                    JObject obj = (JObject)value;
+                    if (definition["path"].ToString() == obj["path"].ToString())
+                    {
+                        definitions.Add(definition);
+                    }
+                }
+            }
+            return definitions;
+        }
+
+        private HttpClient GetClient(string url)
+        {
+            HttpClient client = new HttpClient();
+            Uri uri = new Uri(url); 
+            client.BaseAddress = new Uri(string.Format("{0}://{1}/", uri.Scheme, uri.Authority));
+            client.DefaultRequestHeaders.Clear();
+            client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Basic", _credentials);
+            return client;
```

Comment:
Which package are you referring too.  Other implementations of the API which we are using currently in PipeBuild didn't not allow the same amount of control over build definitions.  The raw REST API gave us the greatest amount of flexibility.  If there is something else which allows us to download, list, update, and create full definitions, then I'm not opposed to investigating them.